File: chapter-03-sites/csharp/SiteOperations.cs
/*
 * SiteOperations.cs
 * Chapter 03: Sites
 *
 * SharePoint Site Operations Example
 * Demonstrates listing, creating, and retrieving SharePoint sites.
 *
 * Uses the shared auth module (common/SharePointGraphAuth): the Graph client is
 * injected via constructor (DI), not built inside the class.
 */

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using LaExperiencia.SharePoint.Common;
using Microsoft.Graph;
using Microsoft.Graph.Models;

namespace LaExperiencia.SharePoint.Chapter03.Sites
{
    /// <summary>
    /// Example class demonstrating SharePoint site operations.
    /// </summary>
    public class SiteOperations
    {
        private readonly GraphServiceClient _graphClient;

        /// <summary>
        /// Creates an instance with an injected Graph client (DI).
        /// </summary>
        public SiteOperations(GraphServiceClient graphClient)
        {
            _graphClient = graphClient ?? throw new ArgumentNullException(nameof(graphClient));
        }

        /// <summary>Lists all sites in the organization.</summary>
        public async Task ListSitesAsync()
        {
            try
            {
                Console.WriteLine("Fetching all sites...");

                var sites = await _graphClient.Sites.GetAsync();

                Console.WriteLine($"\nFound {sites?.Value?.Count ?? 0} sites:");
                Console.WriteLine(new string('-', 80));

                foreach (var site in sites?.Value ?? new List<Site>())
                {
                    Console.WriteLine($"Display Name: {site.DisplayName}");
                    Console.WriteLine($"  ID: {site.Id}");
                    Console.WriteLine($"  Web URL: {site.WebUrl}");
                    Console.WriteLine(new string('-', 80));
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error listing sites: {ex.Message}");
                throw;
            }
        }

        /// <summary>Gets a specific site by hostname and site path (e.g. contoso.sharepoint.com / book-test).</summary>
        public async Task<Site?> GetSiteAsync(string hostname, string sitePath)
        {
            try
            {
                Console.WriteLine($"Fetching site: {hostname}/sites/{sitePath}");

                var site = await _graphClient.Sites[$"{hostname}:/sites/{sitePath}"].GetAsync();

                if (site != null)
                {
                    Console.WriteLine("Site found:");
                    Console.WriteLine($"  Display Name: {site.DisplayName}");
                    Console.WriteLine($"  ID: {site.Id}");
                    Console.WriteLine($"  Web URL: {site.WebUrl}");
                    Console.WriteLine($"  Description: {site.Description}");
                }

                return site;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error getting site: {ex.Message}");
                throw;
            }
        }

        /// <summary>Gets site by its unique identifier.</summary>
        public async Task<Site?> GetSiteByIdAsync(string siteId)
        {
            try
            {
                Console.WriteLine($"Fetching site by ID: {siteId}");

                var site = await _graphClient.Sites[siteId].GetAsync();

                if (site != null)
                {
                    Console.WriteLine("Site found:");
                    Console.WriteLine($"  Display Name: {site.DisplayName}");
                    Console.WriteLine($"  Web URL: {site.WebUrl}");
                }

                return site;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error getting site by ID: {ex.Message}");
                throw;
            }
        }

        /// <summary>Searches for sites by keyword.</summary>
        public async Task SearchSitesAsync(string keyword)
        {
            try
            {
                Console.WriteLine($"Searching for sites with keyword: '{keyword}'");

                var sites = await _graphClient.Sites.GetAsync(requestConfiguration =>
                {
                    requestConfiguration.QueryParameters.Search = keyword;
                });

                Console.WriteLine($"\nFound {sites?.Value?.Count ?? 0} matching sites:");
                Console.WriteLine(new string('-', 80));

                foreach (var site in sites?.Value ?? new List<Site>())
                {
                    Console.WriteLine($"Display Name: {site.DisplayName}");
                    Console.WriteLine($"  Web URL: {site.WebUrl}");
                    Console.WriteLine($"  Description: {site.Description}");
                    Console.WriteLine(new string('-', 80));
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error searching sites: {ex.Message}");
                throw;
            }
        }

        /// <summary>Gets the root site of the organization.</summary>
        public async Task<Site?> GetRootSiteAsync()
        {
            try
            {
                Console.WriteLine("Fetching root site...");

                var site = await _graphClient.Sites["root"].GetAsync();

                if (site != null)
                {
                    Console.WriteLine("Root site:");
                    Console.WriteLine($"  Display Name: {site.DisplayName}");
                    Console.WriteLine($"  ID: {site.Id}");
                    Console.WriteLine($"  Web URL: {site.WebUrl}");
                }

                return site;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error getting root site: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Entry point: builds the Graph client with the common module (client secret) and
        /// runs a read-only demo against the book-test test site.
        /// </summary>
        public static async Task Main(string[] args)
        {
            try
            {
                var graphClient = SharePointGraphClientFactory.CreateFromSecret();
                var siteOps = new SiteOperations(graphClient);

                // Test site (overridable via SHAREPOINT_HOSTNAME / SHAREPOINT_SITE_PATH env vars).
                var hostname = Environment.GetEnvironmentVariable("SHAREPOINT_HOSTNAME") ?? "olddogsoft1.sharepoint.com";
                var sitePath = Environment.GetEnvironmentVariable("SHAREPOINT_SITE_PATH") ?? "book-test";

                await siteOps.GetSiteAsync(hostname, sitePath);
                await siteOps.ListSitesAsync();

                Console.WriteLine("\nSite operations completed successfully!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}

File: chapter-03-sites/java/SiteOperations.java
/*
 * SiteOperations.java
 * Chapter 03: Sites
 *
 * SharePoint Site Operations Example
 * Demonstrates listing, creating, and retrieving SharePoint sites
 */

package com.sharepointexperience.chapter03;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonArray;

/**
 * Site data class
 */
class Site {
    String id;
    String name;
    String displayName;
    String webUrl;
    String description;
}

/**
 * Response wrapper for site list
 */
class SiteListResponse {
    JsonArray value;
}

/**
 * Example class demonstrating SharePoint site operations
 */
public class SiteOperations {
    private final String tenantId;
    private final String clientId;
    private final String clientSecret;
    private final HttpClient httpClient;
    private final Gson gson;
    private String accessToken;

    /**
     * Initializes a new instance of SiteOperations
     */
    public SiteOperations() {
        this.tenantId = System.getenv("TENANT_ID");
        this.clientId = System.getenv("CLIENT_ID");
        this.clientSecret = System.getenv("CLIENT_SECRET");
        this.httpClient = HttpClient.newHttpClient();
        this.gson = new Gson();

        validateConfig();
    }

    private void validateConfig() {
        if (tenantId == null || clientId == null || clientSecret == null) {
            throw new IllegalArgumentException(
                "TENANT_ID, CLIENT_ID, and CLIENT_SECRET environment variables are required"
            );
        }
    }

    /**
     * Gets access token for Microsoft Graph
     */
    private String getAccessToken() throws IOException, InterruptedException {
        if (accessToken != null) return accessToken;

        String tokenEndpoint = String.format(
            "https://login.microsoftonline.com/%s/oauth2/v2.0/token",
            tenantId
        );

        String form = String.format(
            "client_id=%s\u0026client_secret=%s\u0026scope=%s\u0026grant_type=client_credentials",
            clientId, clientSecret, "https://graph.microsoft.com/.default"
        );

        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(tokenEndpoint))
            .header("Content-Type", "application/x-www-form-urlencoded")
            .POST(HttpRequest.BodyPublishers.ofString(form))
            .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 200) {
            throw new IOException("Failed to get access token: " + response.body());
        }

        JsonObject jsonResponse = gson.fromJson(response.body(), JsonObject.class);
        accessToken = jsonResponse.get("access_token").getAsString();
        return accessToken;
    }

    /**
     * Lists all sites in the organization
     */
    public void listSites() throws IOException, InterruptedException {
        try {
            System.out.println("Fetching all sites...\n");

            String token = getAccessToken();
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://graph.microsoft.com/v1.0/sites"))
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/json")
                .GET()
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new IOException("Failed to list sites: " + response.body());
            }

            JsonObject jsonResponse = gson.fromJson(response.body(), JsonObject.class);
            JsonArray sites = jsonResponse.getAsJsonArray("value");

            System.out.println("Found " + sites.size() + " sites:");
            System.out.println("-".repeat(80));

            for (int i = 0; i < sites.size(); i++) {
                JsonObject site = sites.get(i).getAsJsonObject();
                System.out.println("Display Name: " + safeStr(site, "displayName", safeStr(site, "name", "N/A")));
                System.out.println("  ID: " + safeStr(site, "id", "N/A"));
                System.out.println("  Web URL: " + safeStr(site, "webUrl", "N/A"));
                System.out.println("-".repeat(80));
            }

        } catch (Exception e) {
            System.err.println("Error listing sites: " + e.getMessage());
            throw e;
        }
    }

    /** Returns the string value of a key from a JsonObject, or default if null/absent. */
    private String safeStr(JsonObject obj, String key, String defaultValue) {
        if (obj.get(key) != null && !obj.get(key).isJsonNull()) {
            return obj.get(key).getAsString();
        }
        return defaultValue;
    }

    /**
     * Gets the root site of the organization
     */
    public void getRootSite() throws IOException, InterruptedException {
        try {
            System.out.println("Fetching root site...");

            String token = getAccessToken();
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://graph.microsoft.com/v1.0/sites/root"))
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/json")
                .GET()
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new IOException("Failed to get root site: " + response.body());
            }

            JsonObject site = gson.fromJson(response.body(), JsonObject.class);

            System.out.println("\nRoot site:");
            System.out.println("  Display Name: " + safeStr(site, "displayName", safeStr(site, "name", "N/A")));
            System.out.println("  ID: " + safeStr(site, "id", "N/A"));
            System.out.println("  Web URL: " + safeStr(site, "webUrl", "N/A"));

        } catch (Exception e) {
            System.err.println("Error getting root site: " + e.getMessage());
            throw e;
        }
    }

    /**
     * Gets a specific site by ID
     */
    public void getSiteById(String siteId) throws IOException, InterruptedException {
        try {
            System.out.println("Fetching site by ID: " + siteId);

            String token = getAccessToken();
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://graph.microsoft.com/v1.0/sites/" + siteId))
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/json")
                .GET()
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new IOException("Failed to get site: " + response.body());
            }

            JsonObject site = gson.fromJson(response.body(), JsonObject.class);

            System.out.println("\nSite found:");
            System.out.println("  Name: " + site.get("name").getAsString());
            System.out.println("  Web URL: " + site.get("webUrl").getAsString());

        } catch (Exception e) {
            System.err.println("Error getting site by ID: " + e.getMessage());
            throw e;
        }
    }

    /**
     * Entry point for the example
     */
    public static void main(String[] args) {
        try {
            System.out.println("=== SharePoint Site Operations Example ===\n");

            SiteOperations siteOps = new SiteOperations();

            // Get root site
            siteOps.getRootSite();

            // List all sites
            siteOps.listSites();

            System.out.println("\nSite operations completed successfully!");

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}
